### PR TITLE
Semret/Layla - OO-Ride Share Edges

### PR DESCRIPTION
**Pull request by snicodimos**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer | Instructor Feedback
:------------- | :------------- | ---
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | In general we were following the linear progress of the requirement until we noticed that in wave 2 certain tests weren't passing until all the different methods were created. Thus, we decided to create all necessary method/helper methods and then run the rest. We looked at the whole picture of the project vs. wave by wave. 
Describe any examples of composition that you encountered in this project, if any | The User were connected by inheritance Driver, where as Trip and Dispatcher (User/Driver) classes where connected via having distance of objects referring to the same address. For example trip dispatcher points to an array of drivers and the first driver in the index points to an instance of driver that has id, vin status etc. | Your answer gets at this, but I want to call out this vocab explicitly. By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
Describe the relationship between `User` and `Driver`| Driver was a sub-class which inherited from superclass User
Describe a _nominal test_ that you wrote for this assignment. | Some of the newer methods are sequential and relied on each other to work. For instance net expenditures relied on total revenue 's method to properly execute the code.  
Describe an _edge case test_ that you wrote for this assignment | For the initialize method in driver class we did some edge cases like bogus entries and nil status for drivers. If nil we had it assigned to unavailable and if bogus entry it raise an argument error. 
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | The whole concept of inheritance was very clear between driver and user. And the fact that a driver can be a passenger and that it gets all the attributes from the user class. Also the connections between all the classes and how to access each components depending which class we are working from has been made clear with this project. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | Initially we were doing all the work together and swapping between test write and man code writing. We also divided portions of the work and independently worked on it and each pair was responsible to teach the other. The pain point was we had open and good communications that helped the progression of the project.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
+require 'csv'
+require 'pry'
+
+module RideShare
+  class Driver < User
+    attr_reader :vin, :driven_trips
+    attr_accessor :status
+
+    VALID_STATUSES = [:AVAILABLE, :UNAVAILABLE]
+
```
Comment (droberts-sea) [inline lib/driver.rb:10]: Great use of a constant here!

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    def total_revenue
+      if @driven_trips == nil
+        return 0
+      end
+
+      total_revenue = 0.0
+      @driven_trips.each do |trip|
+        x = (trip.cost - 1.65)
+        y = x * 0.08
+        total_revenue += (x - y)
```
Comment (droberts-sea) [inline lib/driver.rb:79]: This code would be a little more readable if the two magic numbers (`0.8` and `1.65`) were stored in constants, maybe something like `DRIVER_CUT` and `FEE`.

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      end
+
+      return average_rating
+    end
+
+    # total revenue calculator
+    def total_revenue
+      if @driven_trips == nil
+        return 0
+      end
```
Comment (droberts-sea) [inline lib/driver.rb:73]: Why would `@driven_trips` ever be `nil`?

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+      @driven_trips << trip
+
+      return @driven_trips
+    end
+
+    def most_recent_trip
+
+      return @driven_trips.sort_by{ |trip| trip.end_time }.first
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:51]: I love this helper method! Good thinking!

File: lib/driver.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+
+    # average rating for drivers
+    def average_rating
+      rating_sum = 0.0
+      @driven_trips.each do |trip|
+        rating_sum += trip.rating
+      end
+      if @driven_trips.length == 0
```
Comment (droberts-sea) [inline lib/driver.rb:60]: Neither of the methods here that work with the list of driven trips account for an incomplete trip.

File: lib/user.rb [hunk 6]
```diff
@@ -16,5 +16,17 @@ def initialize(input)
     def add_trip(trip)
       @trips << trip
     end
+
+    def net_expenditures
+      return   @trips.reduce(0) do  |total_cost, trip|
+        total_cost += trip.cost
+      end
```
Comment (droberts-sea) [inline lib/user.rb:23]: Good use of enumerables here. However, both these methods will hit an error if you try to run them on a user with an incomplete trip. There are a couple of workarounds for this.

You could explicitly ignore trips with a `nil` cost:

```ruby
trips.each do |trip|
  if trip.cost.nil?
    next
  end
  # ... add to the total ...
end
```

Or even better, you could write a helper method that returns a list of complete trips:

```ruby
def completed_trips
  return @trips.reject { |t| t.end_time.nil? }
end

def net_expenditures
  completed_trips.each do |trip|
    # ... same logic as before ...
  end
end
```

File: lib/driver.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      end
+
+      return total_revenue.round(2)
+    end
+
+
+    def net_expenditures
+      net_expenditures = super - self.total_revenue
+      return net_expenditures.round(2)
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:89]: Good use of `super` here!

File: specs/user_spec.rb [hunk 8]
```diff
@@ -58,5 +55,47 @@
         expect(trip.passenger.id).must_equal 9
       end
     end
+
+    it "return total cost for trips per user" do
+
+      @user = RideShare::User.new(id: 8, name: "Merl Glover III",
+        phone: "1-602-620-2330 x3723", trips: [])
```
Comment (droberts-sea) [inline specs/user_spec.rb:62]: I think you're missing some edge cases that are common to both these `User` methods:

- What happens if the user has no trips?
- What happens if the user has an incomplete trip?

File: specs/driver_spec.rb [hunk 9]
```diff
@@ -1,104 +1,194 @@
...
-      expect(average).must_be :<=, 5.0
+    it "total revenue returns a float" do
+      expect(@driver.total_revenue).must_be_kind_of Float
     end
 
-    it "returns zero if no trips" do
+    it " total revenue returns zero if no driven trips" do
       driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV",
-                                     vin: "1C9EVBRM0YBC564DZ")
-      expect(driver.average_rating).must_equal 0
```
Comment (droberts-sea) [inline specs/driver_spec.rb:83]: Good work catching this edge case.

File: lib/trip_dispatcher.rb [hunk 10]
```diff
@@ -60,11 +72,100 @@ def find_passenger(id)
...
+
+      available_driver = select_driver(user_id)
+
+
+      # find { |driver| driver.status == :AVAILABLE && driver.id != passenger.id}
+
+      # make an exception to let the user know there are no available drivers
+      if available_driver == nil
+        raise StandardError.new("There are no avaiable drivers")
+      end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:147]: Instead of using a generic `StandardError`, you should define your own custom exception type.

File: specs/trip_dispatch_spec.rb [hunk 11]
```diff
@@ -76,39 +74,92 @@
...
+      end
+
+      it "is the trip created properly?" do
+        expect(@trip).must_be_instance_of RideShare::Trip
+      end
+
+      it "Drivers cannot drive themselves" do
+        expect(@trip.driver).wont_be_same_as @trip.passenger
+      end
+
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:125]: Is there a line of code you could remove to make this test fail? I suspect that even if you didn't have checks for this in your code, it's likely the same driver would not be assigned.

Instead, I would loop through the entire list of drivers in the trip dispatcher and make all of them except for one be unavailable. Then request a trip for that driver (as a passenger). It should result in an error.

File: specs/trip_dispatch_spec.rb [hunk 12]
```diff
@@ -76,39 +74,92 @@
...
+      it "Was the trip list for trip dispatcher updated?" do
+        # this checks to see if the new trip (@trip) is in the dispatcher's trips
+        expect(@dispatcher.trips.find { |t| t.id == @trip.id }).wont_be_nil
+      end
+
+      it "What happens if you try to request a trip when there are no AVAILABLE drivers?" do
+
+          @dispatcher.drivers.each do |driver|
+            if driver.status == :AVAILABLE
+              driver.status = :UNAVAILABLE
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:149]: I love this test! Pro tip: you don't need to check that the status is `:AVAILABLE` before setting it to `:UNAVAILABLE`, because setting a variable to the value it already holds is a no-op.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes - see responses above
**Wave 1** |
Appropriate use of Ruby's `Date`| yes
`Trip` has a helper method to calculate duration| yes
`User` (passenger) has a method to calculate total cost of all trips| yes
Tests for wave 1 | missing some edge cases - see inline
**Wave 2** |
`Driver` inherits from `User` | yes
`Driver` has `add_driven_trip` method | yes
`Driver` has method to calculate average rating | yes
`Driver` has method to calculate net expenditures and it uses `super` | yes
`Driver` has a method to calculate total revenue | yes
Tests for wave 2 | yes
**Wave 3** |
`TripDispatcher` has a new method to create trips | yes
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | yes
Complex logic was correctly implemented | yes
Tests for `request_trip` | yes
Methods from wave 1 and 2 handle incomplete trips | no - see inline
Tests for wave 1 and 2 methods with incomplete trips | no
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | yes
Appropriate helper methods were made to help with complex logic | yes
Tests for wave 4 | yes
**Overall** |

**Review by droberts-sea — COMMENTED** [review]
Great work overall! There are a few places where code could be cleaned up or test coverage expanded that I've tried to call out inline, but in general I'm pretty happy with this submission. Keep up the hard work!
